### PR TITLE
[SPARK-10816][SS] SessionWindow support for Structure Streaming

### PR DESCRIPTION
**Pull request by viirya**
### What changes were proposed in this pull request?

This patch proposes to add session window feature to Structured Streaming.

The code is based on previous PR #22583. Credits to original authors and they should be listed as major authors in the commit.

Changes:

* Remove unnecessary partial aggregation.
* Configurable state store format: in case we want to use different state format for session window in the next releases. We should let it configurable like aggregation. Current the format config value is 1.
* Replace state store format: session windows per key are not stored as a long array anymore. Instead, we store states to two state stores, one for key -> list of start time, and one for (key, start time) -> state value.

### Why are the changes needed?

Support session window feature in Structure Streaming. This feature is well-known one in streaming processing. Based on previous discussion and new customer requirement, it seems to be a solid reason to have this in Spark.

### Does this PR introduce _any_ user-facing change?

Yes, users may be able to use session window feature in Structure Streaming.

### How was this patch tested?

Unit test currently. We may put it to staging test internally.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/AggUtils.scala [hunk 1]
```diff
@@ -262,6 +262,17 @@ object AggUtils {
    *  - PartialMerge (now there is at most 1 tuple per group)
    *  - StateStoreSave (saves the tuple for the next batch)
    *  - Complete (output the current result of the aggregation)
+   *
+   * Plans a streaming aggregation with Session Window using the following progression:
+   *  - (Shuffle + Session Window Assignment, see `SessionWindowExec`)
+   *  - Partial Aggregation (now there is at most 1 tuple per group)
+   *  - SessionStateStoreRestore (now there is 1 tuple from this batch + optionally one from
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/AggUtils.scala:269]: This process was simplified, compared with original PR.

Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/AggUtils.scala:269]: As it removed unnecessary `PartialMerge`, I think this could improve the performance.

**Comment by viirya** [comment]
cc @xuanyuanking

**Comment by viirya** [comment]
retest this please

**Comment by viirya** [comment]
This approach is intuitive, so the change footprint is relatively smaller. It is also easier to maintain. Per performance wise, it in on par with the alternative one.

I optimized it more from original implementation. There are some I think we can work. E.g. isolating state store format to an manager class, so we may change state store format if we need. Dynamic session window gap, etc. I would not do that now before we reach a consensus if we are all agreed to pursue this approach.

BTW, I may also take time to play with the alternative approach.

cc @dbtsai

**Comment by viirya** [comment]
I'd like to invite some streaming folks to look at this @HeartSaVioR @xuanyuanking @gaborgsomogyi @zsxwing

**Comment by viirya** [comment]
retest this please

**Comment by viirya** [comment]
@xuanyuanking @HeartSaVioR State store format was added. State store format was revised with two stores instead of a long array of all states together. Appreciate if you have time to look at this.

cc @gaborgsomogyi @zsxwing @tdas @Ngone51

**Comment by xuanyuanking** [comment]
I'm reviewing this. Thank you @viirya !!

**Comment by HeartSaVioR** [comment]
Thanks for taking up the missing piece in SS! I'm happy to see we continue the investment for this feature. Let me take a look.

**Review by xuanyuanking — COMMENTED** [review]
Finish a rough round. Will take a detailed look for `StreamingSessionWindowStateManager`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 2]
```diff
@@ -673,6 +673,19 @@ case class Window(
   def windowOutputSet: AttributeSet = AttributeSet(windowExpressions.map(_.toAttribute))
 }
 
+case class SessionWindow(
```
Comment (HeartSaVioR) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala:676]: Ah OK this is conflicting with SessionWindowExpression when renaming. OK to leave as it is.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SessionWindowExec.scala [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+  override def requiredChildOrdering: Seq[Seq[SortOrder]] =
+    Seq(sessionSpec.map(SortOrder(_, Ascending)) :+ SortOrder(timeColumn, Ascending))
+
+  override def producedAttributes: AttributeSet = AttributeSet(windowExpressions.toAttribute)
+
+  override def output: Seq[Attribute] = child.output ++ Seq(windowExpressions.toAttribute)
+
+  override def outputPartitioning: Partitioning = child.outputPartitioning
+
+  override def outputOrdering: Seq[SortOrder] = child.outputOrdering
```
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SessionWindowExec.scala:61]: This made me stop to think, and looks like this is guaranteed per micro-batch. Probably ideal to leave a brief explanation how it is guaranteed.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SessionWindowExec.scala [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+
+  /**
```
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SessionWindowExec.scala:63]: The comment doesn't seem to be needed.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SessionWindowMergeExec.scala [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.catalyst.plans.physical.{ClusteredDistribution, Distribution, Partitioning}
+
+/**
+ * The physical plan for streaming query, merge session window after restore from state store.
+ * Note: the end time of window that restore from statestore has already contain session windowGap
+ *
+ * @param windowExpressions
```
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SessionWindowMergeExec.scala:31]: add explanation on each param, or simply remove

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala [hunk 6]
```diff
@@ -348,7 +348,6 @@ abstract class SparkStrategies extends QueryPlanner[SparkPlan] {
           rewrittenResultExpressions,
           stateVersion,
           planLater(child))
-
```
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala:351]: nit: unnecessary change

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSessionWindowingSuite.scala [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+    try {
+      f(tableName)
+    } finally {
+      spark.catalog.dropTempView(tableName)
+    }
+  }
+
+  test("session window in SQL with single key as session window key") {
+    withTempTable { table =>
+      val a = spark.sql(
```
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameSessionWindowingSuite.scala:50]: dead code?

**Review by HeartSaVioR — COMMENTED** [review]
Still reviewing. Please consider the files what I've not commented are not reviewed yet.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SessionWindowExec.scala [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+   * Produces the result of the query as an `RDD[InternalRow]`
+   *
+   * Overridden by concrete implementations of SparkPlan.
+   */
+  override protected def doExecute(): RDD[InternalRow] = {
```
[the remark this answers is not in the document]
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SessionWindowExec.scala:68]: And if I understand correctly, this needs to copy all input rows because of the buffering. Now I could reload the context of my patch a bit; MergingSessionsIterator only copy the rows which are the first row of session, as it only needs to retain the last session to compare with current input row.

[the remark this answers is not in the document]
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SessionWindowExec.scala:68]: I think it's valid to compare with UpdatingSessionIterator, which is used to support aggregation with one distinct and technically does the same with this. The major difference between twos looks to be that UpdatingSessionIterator doesn't try to memorize entire parts of row - it only memorizes the value part of row, as key part should be just all same with current session, and when session is closed, it restores the memorized rows with key & value parts.

I guess there're pros and cons against twos but not that outstanding, and at least this is simpler, I'm OK to pick this up for the replacement of UpdatingSessionIterator. I'm feeling that MergingSessionsIterator is something we should revisit, but as I said, if we feel that's a blocker on moving forward, I'll take it up after this lands to the codebase.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SessionWindowMergeExec.scala [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+    child: SparkPlan)
+  extends UnaryExecNode {
+
+  override def requiredChildDistribution: Seq[Distribution] = {
+    ClusteredDistribution(sessionSpec) :: Nil
+  }
+
+  // Data should be sorted, so we can merge session window directly.
+  // TODO: use this requirement for simplicity, not necessary to sort the whole dataset,
+  // try better way later.
```
[the remark this answers is not in the document]
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SessionWindowMergeExec.scala:47]: If I remember correctly, we still have to sort to handle new event earlier than session stored in state store. "event time processing" means events can be "out of order".

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/statefulOperators.scala [hunk 10]
```diff
@@ -438,6 +443,434 @@ case class StateStoreSaveExec(
...
+          // after the `getKey` the key = b, the preKey also is b, however, the expected value
+          // of preKey is a
+          preKey = key.copy
+        }
+
+        if (savedState == null) {
+          numOutputRows += 1
+          Seq(row)
+        } else {
+          val outputs = savedState :+ row
```
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/statefulOperators.scala:514]: As I said, I don't think this retains the overall order hence requiring additional sort. `savedState` can be injected anywhere in input rows.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SessionWindowExec.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SessionWindowExec.scala:68]: That's right. It is a trade-off. For session window use case, I image that most cases the buffer should not hold too many rows. It's hardly to think about that a session window has rows causing serious problem. Sounds like a rare case to me.

I think I understand the design thinking here. The approach `MergingSessionsIterator` takes increases complexity by mixing aggregating and session window operations. If don't mix with aggregation, currently I don't have idea how to avoid buffering. I may not worry much too early about this before we have real customer complaints on this issue.

[the remark this answers is not in the document]
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/SessionWindowExec.scala:68]: That is a one of concerns. Another concern is, to buffer row you'll need to "copy" the row, which makes entire input rows being copied. I see there're multiple physical ops to buffer rows, which makes me wondering about the performance and resource usage.

I'll need to check the performance is really on par with mine - I think the major complexity of mine was introduced on linked-list of state format. Migrating state format to the one in agreement here would reduce the complexity significantly, so after applying the change on mine, I think we need to reevaluate both properly.

**Comment by HeartSaVioR** [comment]
I did some performance tests I did before, and I observed outstanding difference between revised my PR (my PR + state format used here) vs this PR.

> revised my PR (versioned as 3.2.0-SPARK-10816-heartsavior)

https://github.com/HeartSaVioR/spark/tree/SPARK-10816-heartsavior-rebase-apply-PR-31570-versioned

> this PR (versioned as 3.2.0-PR-31570)

https://github.com/HeartSaVioR/spark/tree/PR-31570-versioned

> benchmark code

https://github.com/HeartSaVioR/iot-trucking-app-spark-structured-streaming/tree/benchmarking-SPARK-10816

I built the benchmark code against locally installed Spark artifacts for both (that said, I built the benchmark code per each).

Simple, change built.sbt, and run `sbt clean assembly`.

> machine to run benchmark

* AMD Ryzen 5600X (no overclock, 3.7 Ghz to 4.6 Ghz, 6 physical cores, 12 logical cores)
* DDR4 3200Mhz 16 GB * 2
* Ubuntu 20.04

Giving `local[*]` showed instability on performance so fixed the value to 8. There're not many physical cores so I reduced the number of partitions down to 5 as well.

> plenty of rows in session

```
./bin/spark-submit --master "local[8]" --conf spark.sql.shuffle.partitions=5 --driver-memory 16g --class com.hortonworks.spark.benchmark.streaming.sessionwindow.plenty_of_rows_in_session.BenchmarkSessionWindowListenerWordCountSessionFunctionAppendMode ./iot-trucking-app-spark-structured-streaming-<version>.jar --query-status-file /tmp/a.json --rate-row-per-second 200000 --rate-ramp-up-time-second 10
```

[plenty-of-rows-in-session-append-mode-mine-rate-200000-v1.txt](https://github.com/apache/spark/files/6174672/plenty-of-rows-in-session-append-mode-mine-rate-200000-v1.txt)

[plenty-of-rows-in-session-append-mode-PR-31570-rate-200000-v1.txt](https://github.com/apache/spark/files/6174674/plenty-of-rows-in-session-append-mode-PR-31570-rate-200000-v1.txt)

* mine showed 160,000+ on processedRowsPerSecond.
* PR-31570 didn't reach 60,000 on processedRowsPerSecond.

> plenty of keys

```
./bin/spark-submit --master "local[8]" --conf spark.sql.shuffle.partitions=5 --driver-memory 16g --class com.hortonworks.spark.benchmark.streaming.sessionwindow.plenty_of_keys.BenchmarkSessionWindowListenerWordCountSessionFunctionAppendMode ./iot-trucking-app-spark-structured-streaming-<version>.jar --query-status-file /tmp/b.json --rate-row-per-second 12000000 --rate-ramp-up-time-second 10
```

[plenty-of-keys-append-mode-mine-rate-12000000-v1.txt](https://github.com/apache/spark/files/6174671/plenty-of-keys-append-mode-mine-rate-12000000-v1.txt)

[plenty-of-keys-append-mode-PR-31570-rate-12000000-v1.txt](https://github.com/apache/spark/files/6174675/plenty-of-keys-append-mode-PR-31570-rate-12000000-v1.txt)

* mine showed "over" 12,000,000 on processedRowsPerSecond. (Probably could reach more if we increase rate.)
* PR-31570 didn't reach 10,000,000 on processedRowsPerSecond.

It'd be appreciated if anyone in reviewing can take the chance on performance test on their site and update the result. I'd love to see the result objecting my perf test (either my tests with different env/config or new tests), but if no one proves the result objecting mine, I guess we all know we need to make effort on the right direction.

**Comment by viirya** [comment]
Thanks for re-evaluating two approaches. It is valuable.

Basically by leveraging the new state store format, two previous efforts are now pretty close, except for how they handle session merging.

No worry. The precondition to picking the simpler approach, is that two approaches have similar performance. I remember this was claimed in the JIRA. Re-evaluation gives us a different number.

I ran the benchmark locally. Due to the difference of machines, I cannot get the same numbers but I can see there is significant difference between two approaches, i.e., 1) merging then aggregating, 2) merging with aggregating.
 
I think we have a few options.

1. Replace with merging with aggregating (`MergingSessionsIterator`). I'm doing it locally to see if we can get a similar number. It'd be good too @HeartSaVioR would like to create PR against this. So it is easier to incorporate authored commits from all parties. It is also fine if @HeartSaVioR wants to work on it after merging this.
2. Switch to the other previous effort + new state store format.

Either works for me. Actually two options are basically the same logic to me, except for some cosmetic difference.

@xuanyuanking WDYT?

**Comment by HeartSaVioR** [comment]
I can provide a new PR based on mine + proposed state format here. (For sure, I'm willing to add both of you to co-authors.)

Now I'm dealing with update mode (must do) and probably supporting global aggregation on streaming query (optional) which might take a couple of days, but I can raise a PR on WIP and continue working on it as well.

**Comment by xuanyuanking** [comment]
@viirya @HeartSaVioR Agree with both of you. I'm also running the local benchmark. The only concern is on the `MergingSessionsIterator ` detail, it would be great if we can separate the code in the first stage and maybe refractory later.

```
I'm dealing with update mode (required)
```
If it takes too much time, I think the update mode and complete mode (maybe also optional) are good enough.

**Comment by viirya** [comment]
@HeartSaVioR @xuanyuanking Okay. Sounds good to me. I also think append and complete mode might be enough and we can work on update mode later if it takes longer.

**Comment by HeartSaVioR** [comment]
Just submitted mine - #31937
Update mode is addressed. Global aggregation for streaming query is not supported yet, though it's still supported on batch query, so probably wouldn't matter.

**Comment by xuanyuanking** [comment]
Just verified the benchmark. Similar results on my side: The plenty of rows in the session test shows about 2X (70000 vs 37000). I can see many spilling logs for ExternalAppendOnlyMap. Maybe I can do some turning and performance analysis later, will review #31937 first.
